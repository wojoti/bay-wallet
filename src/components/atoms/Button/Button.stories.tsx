import {ComponentMeta, ComponentStory} from '@storybook/react-native';
import {Button} from 'components/atoms';

export default {
  title: 'Atoms/Button',
  component: Button,
  argTypes: {
    onPress: {action: 'onPress'},
    type: {
      control: {type: 'select'},
      options: ['primary', 'white', 'simple', 'alt', 'disabled', 'action'],
    },
  },
} as ComponentMeta<typeof Button>;
const Template: ComponentStory<typeof Button> = args => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: 'Primary button',
  testId: '',
  type: 'primary',
};

export const Secondary = Template.bind({});
Secondary.args = {
  children: 'Secondary button',
  testId: '',
  type: 'secondary',
};
