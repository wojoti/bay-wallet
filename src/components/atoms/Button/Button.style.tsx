import {StyleSheet} from 'react-native';
import {primary, secondary} from 'theme/colors';
import {boldFont} from 'theme/fonts';

interface Props {
  pressed?: boolean;
}

export const textStyle = StyleSheet.create({
  primaryText: {
    margin: 0,
    padding: 0,
    textAlign: 'center',
    fontFamily: boldFont,
    fontSize: 18,
    color: 'white',
  },
  secondaryText: {
    margin: 0,
    padding: 0,
    textAlign: 'center',
    fontFamily: boldFont,
    fontSize: 18,
    color: primary,
  },
  tertiaryText: {
    margin: 0,
    padding: 0,
    textAlign: 'center',
    fontFamily: boldFont,
    fontSize: 18,
    color: 'black',
  },
  tertiaryAlternativeText: {
    margin: 0,
    padding: 0,
    textAlign: 'center',
    fontFamily: boldFont,
    fontSize: 18,
    color: primary,
  },
});

const style = ({pressed}: Props) =>
  StyleSheet.create({
    // wrapper: {
    //   borderWidth: type === 'alt' ? 1 : 0,
    //   elevation: type === 'action' ? 5 : 0,
    //   backgroundColor:
    //     type === 'primary'
    //       ? primary
    //       : type === 'disabled'
    //       ? '#AFAFAF'
    //       : type === 'action'
    //       ? '#4E439E'
    //       : 'white',
    // },
    primaryButton: {
      height: 60,
      borderRadius: 8,
      borderColor: primary,
      borderStyle: 'solid',
      opacity: pressed ? 0.8 : 1,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: primary,
    },
    secondaryButton: {
      height: 60,
      borderRadius: 8,
      borderWidth: 1,
      borderColor: primary,
      borderStyle: 'solid',
      opacity: pressed ? 0.8 : 1,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: secondary,
    },
    tertiaryButton: {
      height: 60,
      borderRadius: 8,
      borderColor: primary,
      borderStyle: 'solid',
      opacity: pressed ? 0.8 : 1,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: secondary,
    },
  });
export default style;
