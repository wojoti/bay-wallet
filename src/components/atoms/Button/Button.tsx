import {PropsWithChildren} from 'react';
import {Pressable, Text} from 'react-native';
import style, {textStyle} from './Button.style';

export type ButtonProps = PropsWithChildren<{
  onPress: () => void;
  type:
    | 'primary'
    | 'secondary'
    | 'tertiary'
    | 'tertiary-alternative'
    | 'disabled';
  testId?: string;
}>;

const Button = ({onPress, testId, type, children}: ButtonProps) => {
  switch (type) {
    case 'primary':
      return (
        <Pressable
          accessibilityRole="button"
          style={({pressed}) => style({pressed}).primaryButton}
          testID={testId}
          onPress={onPress}>
          <Text style={textStyle.primaryText} testID="button-text">
            {children}
          </Text>
        </Pressable>
      );
    case 'secondary':
      return (
        <Pressable
          accessibilityRole="button"
          style={({pressed}) => style({pressed}).secondaryButton}
          testID={testId}
          onPress={onPress}>
          <Text style={textStyle.secondaryText} testID="button-text">
            {children}
          </Text>
        </Pressable>
      );
    case 'tertiary':
      return (
        <Pressable
          accessibilityRole="button"
          style={({pressed}) => style({pressed}).tertiaryButton}
          testID={testId}
          onPress={onPress}>
          <Text style={textStyle.tertiaryText} testID="button-text">
            {children}
          </Text>
        </Pressable>
      );
    case 'tertiary-alternative':
      return (
        <Pressable
          accessibilityRole="button"
          style={({pressed}) => style({pressed}).tertiaryButton}
          testID={testId}
          onPress={onPress}>
          <Text style={textStyle.tertiaryAlternativeText} testID="button-text">
            {children}
          </Text>
        </Pressable>
      );
    case 'disabled':
      return (
        <Pressable
          accessibilityRole="button"
          style={({pressed}) => style({pressed}).primaryButton}
          testID={testId}
          onPress={onPress}
          disabled={true}>
          <Text style={textStyle.primaryText} testID="button-text">
            {children}
          </Text>
        </Pressable>
      );
  }
};

export default Button;
