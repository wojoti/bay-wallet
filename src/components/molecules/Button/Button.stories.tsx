import {ComponentMeta, ComponentStory} from '@storybook/react-native';
import {Button} from 'components/molecules';
import {primary} from 'theme/colors';

export default {
  title: 'Molecules/Button',
  component: Button,
  argTypes: {
    onPress: {action: 'onPress'},
    color: {
      control: {type: 'color'},
    },
    borderColor: {
      control: {type: 'color'},
    },
    fontColor: {
      control: {type: 'color'},
    },
  },
} as ComponentMeta<typeof Button>;
const Template: ComponentStory<typeof Button> = args => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: 'Primary button',
  testId: '',
  height: 65,
  borderRadius: 8,
  borderColor: primary,
  color: primary,
  fontColor: 'white',
};

export const Secondary = Template.bind({});
Secondary.args = {
  children: 'Secondary button',
  testId: '',
  height: 65,
  borderRadius: 8,
  borderColor: primary,
  color: 'white',
  fontColor: primary,
};
