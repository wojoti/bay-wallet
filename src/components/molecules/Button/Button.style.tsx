import {StyleSheet} from 'react-native';
import {primary} from 'theme/colors';
import {boldFont} from 'theme/fonts';

interface PressableProps {
  pressed?: boolean;
  height?: number;
  borderRadius?: number;
  borderColor?: string;
  color?: string;
  fontColor?: string;
}

const style = ({
  pressed,
  height,
  borderRadius,
  borderColor,
  color,
  fontColor,
}: PressableProps) =>
  StyleSheet.create({
    pressable: {
      height: height || 65,
      borderRadius: borderRadius || 8,
      borderColor: borderColor || primary,
      borderStyle: 'solid',
      opacity: pressed ? 0.8 : 1,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: color || primary,
    },
    text: {
      textAlign: 'center',
      fontFamily: boldFont,
      fontSize: 18,
      color: fontColor || 'white',
    },
  });
export default style;
