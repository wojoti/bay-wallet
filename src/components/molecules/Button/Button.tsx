import {PropsWithChildren} from 'react';
import {Text} from 'react-native';
import style from './Button.style';

export type ButtonProps = PropsWithChildren<{
  onPress: () => void;
  height?: number;
  borderRadius?: number;
  borderColor?: string;
  color?: string;
  fontColor?: string;
  testId?: string;
  disabled?: boolean;
}>;

const Button = ({
  onPress,
  height,
  borderRadius,
  borderColor,
  color,
  fontColor,
  disabled = false,
  testId,
  children,
}: ButtonProps) => {
  return (
    <Button
      testId={testId}
      height={height}
      borderRadius={borderRadius}
      borderColor={borderColor}
      color={color}
      onPress={onPress}
      disabled={disabled}>
      <Text
        style={
          style({
            pressed: false,
            height,
            borderRadius,
            borderColor,
            color,
            fontColor,
          }).text
        }
        testID="button-text">
        {children}
      </Text>
    </Button>
  );
};

export default Button;
