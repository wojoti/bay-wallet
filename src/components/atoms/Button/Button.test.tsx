import '@testing-library/jest-native/extend-expect';
import {fireEvent, render, screen} from '@testing-library/react-native';
import TestRenderer from 'react-test-renderer';
import Button, {ButtonProps} from './Button';

const handlePress = jest.fn();
const props: ButtonProps = {
  onPress: handlePress,
  testId: 'test-button-id',
  children: 'test button',
  type: 'primary',
};

const modifiedProps: ButtonProps = {
  ...props,
  type: 'disabled',
};

test('should match snapshot', () => {
  const tree = TestRenderer.create(<Button {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('should render matching button with text', () => {
  render(<Button {...props} />);
  const buttonElement = screen.getByTestId('test-button-id');
  expect(buttonElement).toBeOnTheScreen();
  const textElement = screen.getByText('test button');
  expect(textElement).toBeOnTheScreen();
});

test("shouldn't be able to handle onPress when disabled", () => {
  render(<Button {...modifiedProps} onPress={handlePress} />);
  const buttonElement = screen.getByTestId('test-button-id');
  fireEvent.press(buttonElement);
  expect(buttonElement).toBeDisabled();
  expect(handlePress).toHaveBeenCalledTimes(0);
});

test('should handle onPress event', () => {
  render(<Button {...props} />);
  const buttonElement = screen.getByTestId('test-button-id');
  fireEvent.press(buttonElement);
  expect(buttonElement).toBeEnabled();
  expect(handlePress).toHaveBeenCalledTimes(1);
});
