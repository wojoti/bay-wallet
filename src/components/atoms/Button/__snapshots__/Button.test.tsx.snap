// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#363062",
      "borderColor": "#363062",
      "borderRadius": 8,
      "borderStyle": "solid",
      "height": 60,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
  testID="test-button-id"
>
  <Text
    style={
      {
        "color": "white",
        "fontFamily": "Poppins-Bold",
        "fontSize": 18,
        "margin": 0,
        "padding": 0,
        "textAlign": "center",
      }
    }
    testID="button-text"
  >
    test button
  </Text>
</View>
`;
